flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala

```diff
@@ -0,0 +1,595 @@
+     */
+    def visit(
+        rel: StreamPhysicalRel,
+        requiredTrait: UpdateKindTrait): Option[StreamPhysicalRel] = rel match {
+      case sink: StreamExecSink[_] =>
+        val childModifyKindSet = getModifyKindSet(sink.getInput)
+        val sinkRequiredTraits = sink.sink match {
+          case _: UpsertStreamTableSink[_] =>
+            // support both ONLY_AFTER and BEFORE_AFTER, but prefer ONLY_AFTER
+            Seq(onlyAfterOrNoUpdate(childModifyKindSet), beforeAfterOrNoUpdate(childModifyKindSet))
+          case _: RetractStreamTableSink[_] =>
+            Seq(beforeAfterOrNoUpdate(childModifyKindSet))
+          case _: AppendStreamTableSink[_] | _: StreamTableSink[_] =>
+            Seq(UpdateKindTrait.NO_UPDATE)
+          case ds: DataStreamTableSink[_] =>
+            if (ds.withChangeFlag) {
+              if (ds.needUpdateBefore) {
+                Seq(beforeAfterOrNoUpdate(childModifyKindSet))
+              } else {
+                // support both ONLY_AFTER and BEFORE_AFTER, but prefer ONLY_AFTER
+                Seq(
+                  onlyAfterOrNoUpdate(childModifyKindSet),
+                  beforeAfterOrNoUpdate(childModifyKindSet))
+              }
+            } else {
+              Seq(UpdateKindTrait.NO_UPDATE)
+            }
+        }
+        val children = sinkRequiredTraits.flatMap(t => visitChildren(sink, t))
+        if (children.isEmpty) {
+          None
+        } else {
+          val sinkTrait = sink.getTraitSet.plus(UpdateKindTrait.NO_UPDATE)
+          Some(sink.copy(sinkTrait, children.head).asInstanceOf[StreamPhysicalRel])
+        }
+
+      case _: StreamExecGroupAggregate | _: StreamExecGroupTableAggregate |
+           _: StreamExecLimit =>
+        // Aggregate, TableAggregate and Limit requires update_before if there are updates
+        val requiredChildTrait = beforeAfterOrNoUpdate(getModifyKindSet(rel.getInput(0)))
+        val children = visitChildren(rel, requiredChildTrait)
+        // use requiredTrait as providedTrait, because they should support all kinds of UpdateKind
+        replaceChildrenAndTrait(rel, children, requiredTrait)
+
+      case _: StreamExecGroupWindowAggregate | _: StreamExecGroupWindowTableAggregate |
+           _: StreamExecDeduplicate | _: StreamExecTemporalSort | _: StreamExecMatch |
+           _: StreamExecOverAggregate | _: StreamExecWindowJoin =>
+        // WindowAggregate, WindowTableAggregate, Deduplicate, TemporalSort, CEP, OverAggregate
+        // and WindowJoin require no updates in input
+        val children = visitChildren(rel, UpdateKindTrait.NO_UPDATE)
+        replaceChildrenAndTrait(rel, children, requiredTrait)
+
+      case rank: StreamExecRank =>
+        val rankStrategies = RankProcessStrategy.analyzeRankProcessStrategies(
+          rank.getInput, rank.partitionKey, rank.orderKey)
+        // creates new Rank nodes for every applied RankStrategy
+        val newRankNodes = rankStrategies.map(rank.copy)
```

Comment:
We can't move it into `visitRankStrategies`, because `rank.copy(RankProcessStrategy)` is not a common interface of `StreamExecRank` and `StreamExecSortLimit`. I'm not sure whether we should introduce a common abstract for `StreamExecRank` and `StreamExecSortLimit`.

Reply:
I added a `applyRankStrategy: RankProcessStrategy => StreamPhysicalRel` paramter to the method `visitRankStrategies` to make it possible move rank.copy into this method.